Adds support for visual selection by text object.

ionutvmi opened the pull request:
This adds support for things like `viw`, `vi'`, `vi[`

aioutecism commented on src/Actions/Selection.ts:
```diff
@@ -138,6 +142,37 @@ export class ActionSelection {
         return Promise.resolve(true);
     }
 
+    @PrototypeReflect.metadata(SymbolMetadata.Action.isChange, true)
```
This is only needed when action will make changes to content.

aioutecism commented on src/Actions/Selection.ts:
```diff
@@ -138,6 +142,37 @@ export class ActionSelection {
+        activeTextEditor.selections.forEach(selection => {
+            const match = args.textObject.apply(selection.active);
+            if (match) {
+                ranges.push(match);
```
We should expand current selections rather than replace them.
Something like this should be done:
```ts
ranges.push(match ? selection.union(match) : selection);
```

aioutecism commented on src/Actions/Selection.ts:
```diff
@@ -138,6 +142,37 @@ export class ActionSelection {
+        let ranges: Range[] = [];
+
+        activeTextEditor.selections.forEach(selection => {
+            const match = args.textObject.apply(selection.active);
```
Things are a little wired when the active cursor is at the start of the selection.

aioutecism commented on src/Actions/Selection.ts:
```diff
@@ -138,6 +142,37 @@ export class ActionSelection {
+
+        if (ranges.length === 0) {
+            return Promise.reject<boolean>(false);
+        } else {
```
We keep `else` on a new line in this project.

aioutecism commented on src/Actions/Selection.ts:
```diff
@@ -138,6 +142,37 @@ export class ActionSelection {
     }
 
+    @PrototypeReflect.metadata(SymbolMetadata.Action.isChange, true)
+    static byTextObject(args: {
```
`expandByTextObject` should be more appropriate.

aioutecism commented:
Thank you very much for your contribution!
Cloud you add some tests to the new feature?

ionutvmi commented:
Hi @aioutecism ,

I have addressed the comments.
Not sure how to approach the tests. I did a short experiment. 
Can you please provide some guidelines there ?

aioutecism commented:
@ionutvmi 
Thanks!
The test you added is good! Some more test cases will be better!

The behavior is wired when active cursor is at start of selection. I'll address this in another p-r.
Thanks for the contribution!

aioutecism commented:
@ionutvmi 
fyi: I've done some fix to this feature here: #121.
